src/cli/index.js

```diff
@@ -234,68 +235,129 @@ export function main({
     });
   };
 
-  //
   const runEventuallyWithNetwork = (mutexPort: ?string): Promise<void> => {
-    return new Promise(resolve => {
+    return new Promise((resolve, reject) => {
       const connectionOptions = {
         port: +mutexPort || constants.SINGLE_INSTANCE_PORT,
       };
 
-      const clients = new Set();
-      const server = net.createServer();
+      function startServer() {
+        const clients = new Set();
+        const server = http.createServer(manager);
 
-      server.on('error', () => {
-        // another Yarn instance exists, let's connect to it to know when it dies.
-        reporter.warn(reporter.lang('waitingInstance'));
-        const socket = net.createConnection(connectionOptions);
+        // The server must not prevent us from exiting
+        server.unref();
 
-        socket
-          .on('connect', () => {
-            // Allow the program to exit if this is the only active server in the event system.
-            socket.unref();
-          })
-          .on('close', (hadError?: boolean) => {
-            // the `close` event gets always called after the `error` event
-            if (!hadError) {
-              process.nextTick(() => {
-                resolve(runEventuallyWithNetwork(mutexPort));
-              });
-            }
-          })
-          .on('error', () => {
-            // No server to listen to ? Let's retry to become the next server then.
-            process.nextTick(() => {
-              resolve(runEventuallyWithNetwork(mutexPort));
-            });
+        // No socket must timeout
```

Comment:
`// We are going to hold all connections until this current process is finished so disable timeouts`